rest_framework_json_api/views.py

```diff
@@ -281,11 +281,14 @@ def get(self, request, *args, **kwargs):
 
     def remove_relationships(self, instance_manager, field):
         field_object = getattr(instance_manager, field)
+        class_name = instance_manager.__class__.__name__
 
         if field_object.null:
             for obj in instance_manager.all():
                 setattr(obj, field_object.name, None)
                 obj.save()
+        elif class_name == 'ManyRelatedManager':
```

Comment:
I guess you have this snippet from further below in the code. Actually there are attributes to figure out when a foreign key is a to many relation. See https://docs.djangoproject.com/en/3.0/ref/models/fields/#attributes-for-fields-with-relations

I think `clear` is certainly the right way to go but why is the `else: instance_manager.all().delete()` still needed?

Also I have a feeling the `remove_relationships` makes things more complicated and if following if/else clause would be moved into the method things can be made easier (e.g. no double checking for to many relationship).

```
            # for to one
            if hasattr(related_instance_or_manager, "field"):
                related_instance_or_manager = self.remove_relationships(
                    instance_manager=related_instance_or_manager, field="field")
            # for to many
            else:
                related_instance_or_manager = self.remove_relationships(
                    instance_manager=related_instance_or_manager, field="target_field")
```

Reply:
> why is the else: instance_manager.all().delete() still needed?

In the class [`ManyRelatedManager`](https://github.com/django/django/blob/a34cb5a6d408203f4fbdb364fc9768c026eda224/django/db/models/fields/related_descriptors.py#L970), `.remove()` and `.clear()` are always available. However, in [`RelatedManager`](https://github.com/django/django/blob/a34cb5a6d408203f4fbdb364fc9768c026eda224/django/db/models/fields/related_descriptors.py#L692) they are only available when the fk references are nullable.

Running `pytest -sk test_patch_one_to_many_relationship` without falling back to `.all().delete()` for `RelatedManager` fails because of this: (The project has a great test suite which caught this for me. 🙂)

<img width="569" alt="image" src="https://user-images.githubusercontent.com/10276811/83329931-a5873480-a27b-11ea-8cc4-1cc1b596d3e5.png">

---

> Actually there are attributes to figure out when a foreign key is a to many relation.

Thanks, I considered using this but I am inclined to continue using the class name (or perhaps instanceof if that'd make it better) since we are literally depending on behavior that's changing based on the class (as referenced above). I am open to changing this if you feel strongly about this.

Reply:
Thanks for elaborating. Makes sense.

In terms of `ManyRelatedManager` - this class is created on the fly so a general `instanceof` is not really possible as I see it. Someone even when unlikely could actually replace this class in its code with a different name or even Django itself in a newer version could have a different class name as the name `ManyRelatedManager` doesn't seem to be documented in the documentation. But the foreign key attributes are so if those are being renamed need to deprecated and noticed therefore.
If those attributes are accessed and don't exist anymore an exception would be raised instead of the instances actually being deleted (as this PR actually tries to avoid). 

That's the reasoning behind it that I think if at all possible it would be better to use those attributes unless you see another safe option.

Reply:
> That's the reasoning behind it that I think if at all possible it would be better to use those attributes unless you see another safe option.

Thanks for the explanation, I understood your point now. I have made changes to just check if `.clear()` exists or not instead of looking at the class name.